Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean

```diff
@@ -0,0 +1,415 @@
+    -- and we record that these representatives, when mapped back into the relevant colimits,
+    -- are actually the components of `x`.
+    have e : ∀ j,
+        colimit.ι ((curry.obj F).obj j) (k j) (y j) = limit.π (curry.obj F ⋙ Limits.colim) j x :=
+      fun j => (z j).choose_spec.choose_spec
+    clear_value k y
+    -- A little tidying up of things we no longer need.
+    clear z
+    -- As a first step, we use that `K` is filtered to pick some point `k' : K` above all the `k j`
+    let k' : K := IsFiltered.sup (Finset.univ.image k) ∅
+    -- and name the morphisms as `g j : k j ⟶ k'`.
+    have g : ∀ j, k j ⟶ k' := fun j => IsFiltered.toSup (Finset.univ.image k) ∅ (by simp)
+    clear_value k'
+    -- Recalling that the components of `x`, which are indexed by `j : J`, are "coherent",
+    -- in other words preserved by morphisms in the `J` direction,
+    -- we see that for any morphism `f : j ⟶ j'` in `J`,
+    -- the images of `y j` and `y j'`, when mapped to `F.obj (j', k')` respectively by
+    -- `(f, g j)` and `(𝟙 j', g j')`, both represent the same element in the colimit.
+    have w :
+      ∀ {j j' : J} (f : j ⟶ j'),
+        colimit.ι ((curry.obj F).obj j') k' (F.map ((𝟙 j', g j') : (j', k j') ⟶ (j', k')) (y j')) =
+          colimit.ι ((curry.obj F).obj j') k' (F.map ((f, g j) : (j, k j) ⟶ (j', k')) (y j)) := by
+      intro j j' f
+      have t : (f, g j) =
+          (((f, 𝟙 (k j)) : (j, k j) ⟶ (j', k j)) ≫ (𝟙 j', g j) : (j, k j) ⟶ (j', k')) := by
+        simp only [id_comp, comp_id, prod_comp]
+      erw [Colimit.w_apply', t, FunctorToTypes.map_comp_apply, Colimit.w_apply', e, ←
+        Limit.w_apply' f, ← e]
+      simp
+    -- Because `K` is filtered, we can restate this as saying that
+    -- for each such `f`, there is some place to the right of `k'`
+    -- where these images of `y j` and `y j'` become equal.
+    simp_rw [colimit_eq_iff.{v, v}] at w
+    -- We take a moment to restate `w` more conveniently.
+    let kf : ∀ {j j'} (_ : j ⟶ j'), K := fun {_} {_} f => (w f).choose
+    let gf : ∀ {j j'} (f : j ⟶ j'), k' ⟶ kf f := fun {_} {_} f => (w f).choose_spec.choose
+    let hf : ∀ {j j'} (f : j ⟶ j'), k' ⟶ kf f := fun {_} {_} f =>
+      (w f).choose_spec.choose_spec.choose
+    have wf :
+      ∀ {j j'} (f : j ⟶ j'),
+        F.map ((𝟙 j', g j' ≫ gf f) : (j', k j') ⟶ (j', kf f)) (y j') =
+          F.map ((f, g j ≫ hf f) : (j, k j) ⟶ (j', kf f)) (y j) :=
+      fun {j} {j'} f => by
+      have q :
+        ((curry.obj F).obj j').map (gf f) (F.map _ (y j')) =
+          ((curry.obj F).obj j').map (hf f) (F.map _ (y j)) :=
+        (w f).choose_spec.choose_spec.choose_spec
+      dsimp [curry_obj_obj_obj, curry_obj_obj_map] at q
+      -- porting note: Lean 4 `dsimp` is unfolding `gf` and `hf` :-(
+      -- We fold them back up.
+      -- **TODO** minimise and ask why this is happening
+      change F.map ((𝟙 j', gf f) : (j', k') ⟶ (j', kf f))
```

Comment:
I asked [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/.60dsimp.60.20unfolding.20local.20lets/near/352671008) about the random unfolding.

Reply:
Great! I agree it's not ideal. If you add a link from here to the zulip discussion I'm happy to merge this part as is.

Reply:
The link is there.